pkg/controller/podautoscaler/horizontal.go

```diff
@@ -678,7 +690,30 @@ func (a *HorizontalController) recordInitialRecommendation(currentReplicas int32
 	}
 }
 
-func (a *HorizontalController) reconcileAutoscaler(ctx context.Context, hpaShared *autoscalingv2.HorizontalPodAutoscaler, key string) error {
+var (
+	// errSpec is used to determine if the error comes from the spec of HPA object in reconcileAutoscaler.
+	// All such errors should have this error as a root error so that the upstream function can understand they're spec errors.
+	// e.g., fmt.Errorf("invalid spec%w", errSpec)
+	errSpec error = errors.New("")
+)
+
+func (a *HorizontalController) reconcileAutoscaler(ctx context.Context, hpaShared *autoscalingv2.HorizontalPodAutoscaler, key string) (retErr error) {
+	// reconciliationAction is used to report which actions this reconciliation has taken.
+	reconciliationAction := metrics.ReconciliationActionNone
+	start := time.Now()
+	defer func() {
+		reconciliationError := metrics.ReconciliationErrorNone
+		if retErr != nil {
+			// In case of error, set "internal" as default.
+			reconciliationError = metrics.ReconciliationErrorInternal
+		}
+		if errors.Is(retErr, errSpec) {
+			reconciliationError = metrics.ReconciliationErrorSpec
+		}
+
+		a.metricsRecorder.ObserveReconciliationResult(reconciliationAction, reconciliationError, time.Since(start).Seconds())
```

Comment:
nit: `ObserveReconciliationResult` could accept `time.Duration` as the last argument and transform into seconds on its own. The fact that the prometheus histogram operates on seconds seems like an implementation detail.